Automattic/jetpack
# tools: Improve check-intra-monorepo-deps version handling

#### Changes proposed in this Pull Request:

The old check was kind of dumb: if the existing dep started with `^`
then it would do like "^1.2", otherwise it would do like "1.2.x-dev".
Except for packages, where it would always do like "^1.2".

That broke when used on a plugin release branch where someone had pinned
a dep like "1.2.3".

The new behavior is a bit more accurate: for plugins, it it ends with
"-dev" then it does like "1.2.x-dev", otherwise it does like "^1.2.3"
(matching the recently-changed behavior of tools/version-packages.sh).
For packages it always does like "^1.2", and for everything else it
always does like "1.2.x-dev".

Plus, to remove the need for a pinned "1.2.3" dep, when doing the
composer.lock update it will force a "^1.2.3" dep to install that
specific version rather than allowing a newer version to be pulled in.
With one caveat: indirect deps might still be updated incorrectly.
Humans will have to watch for that, at least until a different PR adds a
CI test for it.

#### Other information:


#### Jetpack product discussion


p1662483262626649-slack-C01U2KGS2PQ

#### Does this pull request change what data or activity we track or use?


No

#### Testing instructions:






* Check out c405506d5b6d1e4951d99aa188b918b4bb0f1448, cherry-pick this into it (`git cherry-pick 7859236 ba6d621 f84d97c`), then run `tools/changelogger-release.sh packages/publicize -p`.
  * [x] The dep on my-jetpack in `projects/plugins/jetpack/composer.json` should be changed to "^2.0.3".
  * [x] Many other deps should change from "^1.2" to "^1.2.3" style.
  * [x] The change in `projects/plugins/jetpack/composer.lock` is only to the record for the publicize package. No other packages change.
* Change PHP package deps in something in `projects/packages/` to "1.2.x-dev" style.
  * [x] Run `tools/check-intra-monorepo-deps.sh`. It should report a proper error.
  * [x] Run `tools/check-intra-monorepo-deps.sh -u`. It should fix it.
* Change PHP package deps in something in `projects/js-packages/` to "^1.2" style.
  * [x] Run `tools/check-intra-monorepo-deps.sh`. It should report a proper error.
  * [x] Run `tools/check-intra-monorepo-deps.sh -u`. It should fix it.
* Change PHP package deps in a plugin to an incorrect version, either `1.2.x-dev`, `^1.2`, `^1.2.3`, or `1.2.3` style.
  * [ ] Run `tools/check-intra-monorepo-deps.sh`. It should report a proper error, suggesting both `1.2.x-dev` and `^1.2.3` styles.
  * [ ] Run `tools/check-intra-monorepo-deps.sh -u`. It should fix it to the corresponding style.
* Try anything else you can think of to break it.
